examples/templates/java/README.md

```diff
@@ -0,0 +1,252 @@
+- Single or multiple Apache Kafka bootstrap servers
+- Apache Kafka SASL/SCRAM authentication over plaintext or SSL connection
+- Secrets vault service [HashiCorp Vault](https://www.vaultproject.io/).
+
+Supported destination configuration:
+- Single Google Pub/Sub topic.
+
+In a simple scenario, the template will create an Apache Beam pipeline that will read messages from a source Kafka server with a source topic, and stream the text messages into specified Pub/Sub destination topic. Other scenarios may need Kafka SASL/SCRAM authentication, that can be performed over plain text or SSL encrypted connection. The template supports using a single Kafka user account to authenticate in the provided source Kafka servers and topics. To support SASL authenticaton over SSL the template will need an SSL certificate location and access to a secrets vault service with Kafka username and password, currently supporting HashiCorp Vault.
+
+## Requirements
+
+- Java 11
+- Kafka Bootstrap Server(s) up and running
+- Existing source Kafka topic(s)
+- An existing Pub/Sub destination output topic
+- (Optional) An existing HashiCorp Vault
+- (Optional) A configured secure SSL connection for Kafka
+
+## Getting Started
+
+This section describes what is needed to get the template up and running.
+- Assembling the Uber-JAR
+- Local execution
+- Google Dataflow Template
+  - Set up the environment
+  - Creating the Dataflow Flex Template
+  - Create a Dataflow job to ingest data using the template.
+- Avro format transferring.
+
+## Assembling the Uber-JAR
+
+To run this template the template Java project should be built into
+an Uber JAR file.
+
+Navigate to the Beam folder:
+
+```
+cd /path/to/beam
+```
+
+In order to create Uber JAR with Gradle, [Shadow plugin](https://github.com/johnrengelman/shadow)
+is used. It creates the `shadowJar` task that builds the Uber JAR:
+
+```
+./gradlew -p examples/templates/java/kafka-to-pubsub clean shadowJar
+```
+
+ℹ️ An **Uber JAR** - also known as **fat JAR** - is a single JAR file that contains
+both target package *and* all its dependencies.
+
+The result of the `shadowJar` task execution is a `.jar` file that is generated
+under the `build/libs/` folder in kafka-to-pubsub directory.
+
+## Local execution
+To execute this pipeline locally, specify the parameters:
+- Kafka Bootstrap servers
+- Kafka input topics
+- Pub/Sub output topic
+in the following format:
+```bash
+--bootstrapServers=host:port \
+--inputTopics=your-input-topic \
+--outputTopic=projects/your-project-id/topics/your-topic-pame
+```
+Optionally, to retrieve Kafka credentials for SASL/SCRAM,
+specify a URL to the credentials in HashiCorp Vault and the vault access token:
+```bash
+--secretStoreUrl=http(s)://host:port/path/to/credentials
+--vaultToken=your-token
+```
+Optionally, to configure secure SSL connection between the Beam pipeline and Kafka,
+specify the parameters:
+- A local path to a truststore file
+- A local path to a keystore file
```

Comment:
Does this work with a local path? Wouldn't the keystore need to get staged for use on a distributed runner?

FWIW I think that configuring KafkaIO to use SSL is much more difficult than it should be. Here's an SO question that describes how you can create a custom ConsumerFactoryFn that downloads a keystore from GCS at execution time: https://stackoverflow.com/questions/42726011/truststore-and-google-cloud-dataflow
I think it could be worthwhile to make KafkaIO do this by default

Reply:
Yes, you're absolutely right, we've made this SSL configuration just as the example for local execution. We updated the code and gave a possibility to store a keystore in GCS and download from it at execution time. If the path starts with `gs://`, then we try to retrieve from GCS, and from the local file otherwise.